### ISSUE TITLE
Room retention *increases* size of state tables.

### ISSUE THREAD

**Issue by matrixbot** [element-hq/synapse#9406]
This issue has been migrated from [#9406](https://github.com/matrix-org/synapse/issues/9406).

---

We have reports that enabling room retention in a very large room (with frequent joins and leaves) actually causes the size of the state tables to *increase*, which is surprising. 

Note that it appears that postgres is vacuuming the tables.

Also note that this is an experimental feature.

**Comment by kaiyou** [comment]
Oh, the original report is pretty old already, no idea why I did not find it in the first place, but I surely was not the original reporter since we experienced this just weeks ago.

**Comment by kaiyou** [comment]
Sorry about spamming this issue. After a first pessimistic report, it appears that rooms that were "de-deduplicated" this time are different from the first time.

My take on this is the retention job did not finish last time, since the database crashed before it could. We were lucky that remaining rooms did not nuke our server once more. Symptoms are similar and the same pattern emerges, this time with even more entries: one room show 180M rows in `state_groups_state` :)

I believe we are now out of most trouble, yet I will keep a copy of the database in case I can help pinpointing the root cause.

**Comment by kaiyou** [comment]
A couple months later, this happened again on our HS, we were lucky to avoid a major downtime, so I went back down the rabbit hole of the history purging code in synapse. Here is a summary of what I found, and how I suggest we fix some of it.

## The process of locating state groups to be deleted is complex and looks flawed

First, history purging happens at : 

https://github.com/element-hq/synapse/blob/5624c8b961ed6a8310a2c6723ae13e854721756b/synapse/storage/databases/main/purge_events.py#L67

This is supposed to purge old events and their relationships, then return every state group to be considered first for deletion. It does this by returning every state group related to any event, including state events, that matches purge criteria, except for outliers. It also ends by removing the reference to state groups from state events matching the criteria and making them outliers. This means those same state events will never be considered in later purges (which explicitly exclude outliers).

I am unsure this is an actual issue, since associated state groups *should* be deleted, and we might not care about these later, but I would feel more confident if we included more state groups as *candidates* than leaving them potentially forever. Maybe this is also related to accumulating unreferenced state groups.

Later on, this initial list is expanded and filtered here:

https://github.com/element-hq/synapse/blob/5624c8b961ed6a8310a2c6723ae13e854721756b/synapse/storage/controllers/purge_events.py#L75

Which basically climbs up the state group tree and stops when finding state groups that are still referenced and thus should not be deleted. This is fine, except it climbs up the tree using `get_previous_state_groups`:

https://github.com/element-hq/synapse/blob/5624c8b961ed6a8310a2c6723ae13e854721756b/synapse/storage/databases/state/store.py#L816

From a quick overview, this does not return previous state groups but filters the state group list, keeping those that have children. This in turn breaks the code from `_find_unreferenced_groups` in a way that simply does not climb up the tree, leaving additional unreferenced groups, possibly forever.

Here is a reference to the separate issue about unreferenced state groups if my analysis can help there: https://github.com/matrix-org/synapse/issues/12821

I suggest:
- fixing the bug in `get_previous_state_groups`
- at least add some documentation to purging code explaining why we dereference state groups, set state events as outliers, etc.

## State groups are stored indefinitely for rejected client events

When creating client events, synapse stores them very early in an `UnpersistedEventContext`. This is done most notably by `calculate_context_info`, which stores references to the event state, thus creating state groups, before the event is checked for authorization, long before it is persisted.

For instance, a client event for joining an invite-only room will generate a state group, even if it is rejected. This leads to unreferenced entries in the state group tree, which also reference unpersisted events.

I am unsure how to proceed about this. On paper, state groups are merely a structured cache, so it should do no harm except for taking up space that is never reclaimed, while changing this behavior sounds like a colossal task.

## Application services weaponize this

At least, Mautrix does.

https://github.com/mautrix/go/blob/57f6cd89e3bf148e7ab497ef076e054fd506e2ab/appservice/intent.go#L91

For every bridged user, instead of checking the room permissions, Mautrix simply tries to join, then upon failure invites the user and joins again. It works fine. Except synapse stores an unreferenced state group in the tree for every bridged user, due to the denied initial join.

A bridged room state group tree is pretty straightforward, except for every user being mentioned three times: one invite, one join, and one unreferenced branch. Due to being unreferenced, it is not considered for deletion when purging the room. However, when removing the main branch, all those state groups are de-delta-ed.

https://github.com/element-hq/synapse/blob/5d3850b0384c7da286031b169d9a37320b6783ea/synapse/storage/databases/state/store.py#L778

So after a history purge, a bridged room has one unreferenced de-delta-ed state group for every room member. This grows quadratically as exposed above, and large rooms may create dozens or hundreds of millions of state group lines in a couple of minutes when purging history due to this, effectively taking the database down.

I suggest:
- for now, stop cleaning state groups when purging history, this leaves unreferenced entries for cleaning scripts to scrub, but I would rather have rotting data than exploding data,
- review the root causes above, as this is hardly an issue with Mautrix and could be exploited in many other ways, including intentionally.

**Comment by kaiyou** [comment]
Reporting back on this issue, this kept happening to us and spamming our production database, so we disabled any purging logics from our main homeserver during the summer, until we took the time to devise a reasonable fix.

In order to fix the issue, I actually tried and went the easy way: we are facing an issue with cached state groups that are unreferenced by any event (because none is persisted), and which *do* reference a previous state group, leading to them being de-delta-ed when the parent state group is deleted.

I figured fixing the issue right in the room purging code, and deleting these obviously state state groups before they cause any damage was sound. It prevents the issue, provides a decent mechanism to regularly spot these nasty groups and remove some rotting data without scanning the entire database (which we might still do in a separate maintenance job), and most of all *does not* require I play around with the even persistence code, which looked daunting.

We will be testing this on our homeserver for the next couple of days, then I'll report back again and provide a PR. Until then, you'll find our work at: https://forge.tedomum.net/tedomum/synapse/-/compare/b066b3aa04248260bd0cfb2ac3569a6e4a3236ea...fix-exploding-state-groups?from_project_id=97&straight=false

**Comment by kaiyou** [comment]
Reporting back one more time after a week of running the "fix". It seems my understanding was incomplete, or my code has issues, because the bug manifested again last night and took down our homeserver for a couple hours after filling the disk with state group lines. I am reviewing the purging code again, and it looks like it is still missing some comments, I am unsure about some of the logic. I am however building an intuition that cleaning state groups just isn't worth it.

- On one hand, I understand that if we are to clean them, we need to do so when purging, because there is no relevant test to later detect unreferenced state groups with certainty, due to state groups being inserted before events are persisted to database.

- On the other hand, I understand that except for a couple special cases with outlier events (which I am unsure about), state events are not deleted when purging, and most state groups are untouched anyway. At best we are saving a couple state group lines every time we purge history, at the cost of one of the most complicated bits of synapse, which does backfire at us due to the even more complicated de-delta-ing part. This is all in favor of a more radical decision: let's stop deleting state groups when we purge history, those are rather cheap anyway, grow almost linearly with the number of state events, and compress well using the state-compressor.

This is however a drastic, no-return decision, and I am unable to weigh things properly.

@erikjohnston sorry for mentioning you, from my lurking on synapse dev channels and Github issues, it sounds like you are the most knowledgeable on the matter. Do you think we should keep fixing the purging code and properly delete state groups while avoiding edge cases that potentially insert millions of new lines? or should we just simplify things and stop removing state groups when we purge history?

Either way, I will keep working on this the next couple of weeks, and will provide a PR as soon as I have tested a working solution.

**Comment by erikjohnston** [comment]
Hey, this is actually something we've been meaning to deal with lately, though haven't been able to carve out the time just yet. Let me brain dump my working understanding of the issue.

There are some quite complex interactions between deleting history, state groups, and event persistence. The major blocking issue with fixing the current code is that there is actually a nasty race when figuring out whether a state group is unreferenced or not: when we persist new state groups they start off as unreferenced and *then* we persist events that reference them. Changing the deleting state group logic to also delete unreferenced state groups runs a much greater risk of accidentally deleting state groups that are "in-flight" (i.e. currently unreferenced but are about to be). I have some thoughts on how we could delete state groups safely, but its non-trivial.

Once we can safely delete state groups, then I think checking if a state group that is about to be de-deltaed is actually referenced *should* fix this issue.

One thing that we have seen repeatedly is that using https://github.com/erikjohnston/synapse-find-unreferenced-state-groups and then deleting the matching state groups reclaims a lot of disk space. (I wouldn't necessarily recommend doing so though)

> On the other hand, I understand that except for a couple special cases with outlier events (which I am unsure about), state events are not deleted when purging, and most state groups are untouched anyway. At best we are saving a couple state group lines every time we purge history, at the cost of one of the most complicated bits of synapse, which does backfire at us due to the even more complicated de-delta-ing part. This is all in favor of a more radical decision: let's stop deleting state groups when we purge history, those are rather cheap anyway, grow almost linearly with the number of state events, and compress well using the state-compressor.

It very much depends on the size and complexity of the room. The state compressor helps a lot, but state can still be one of the major uses of disk space. Certainly we have found that deleting state groups in large rooms does lead to significant disk reduction. The reason for this is that while the state events themselves aren't deleted, we can delete the mapping of event ID to state at that event (which is what state groups actually are).

**Comment by richvdh** [comment]
This is basically the same as https://github.com/matrix-org/synapse/issues/3364#issuecomment-395263702.

**Comment by kaiyou** [comment]
Thanks to both of you for the replies.

> The major blocking issue with fixing the current code is that there is actually a nasty race when figuring out whether a state group is unreferenced or not: when we persist new state groups they start off as unreferenced and then we persist events that reference them.

This is indeed what's preventing from properly deleting these stale state groups, because we can never be certain they are in fact unreferenced.

> One thing that we have seen repeatedly is that using https://github.com/erikjohnston/synapse-find-unreferenced-state-groups and then deleting the matching state groups reclaims a lot of disk space. (I wouldn't necessarily recommend doing so though)

We are indeed running this very specifically for the worst rooms as a last resort.

> Once we can safely delete state groups, then I think checking if a state group that is about to be de-deltaed is actually referenced should fix this issue.

I have been thinking about this a lot lately, and I want to suggest another way to fix things. From my understanding, the issue is twofold: first there are unreferenced state groups in the database, which we can never properly delete, and which sometimes make up such a structure that they get massively de-delta-ed (like here, when an isolated unreferenced state group is generated for every new user joining a Mautrix bridge room) 2. the purge code, and pretty much any code trying to maintain and delete state groups is very complex, making things hard to reason about and patch.

We definitely need to fix the race between persisting state groups and the events that generated them, possibly by changing the core logics of event persistence, or using database transactions for instance. Once this (admittedly huge work) is done, nothing stops us from switching to a GC for deleting state groups.

Switching to garabage collecting would allow us to keep the state group deletion code simpler (just deleted unreferenced state groups instead of extrapolating from the list of purged events for instance) and closer to the rest of state group related code. We could remove anything related to state groups from the room purging code, and maybe explicitely trigger the GC after a room or room history is purged.

In my mental model, state groups are a persistent cache of state history. I think that managing a cache data structure with a GC does make sense. Would you agree with my analysis? and in that case, would you agree we work on the race first, then switch to a GC model? If so, I will focus my effort on helping with the race first, making sure we only persist referenced state groups.

**Comment by erikjohnston** [comment]
This is hopefully fixed by https://github.com/element-hq/synapse/pull/18131 (and co), and is in the v1.124.0 RCs.
